aneekee/nodejs-aws-cart-api
# task-8

Frontend -- https://d1vdv7tmd1pu8.cloudfront.net/
Frontend PR -- https://github.com/slnchn/nodejs-aws-shop-react/pull/6
Lambda URL -- https://ejbywu1m1l.execute-api.eu-west-1.amazonaws.com/prod

## How to test?

Please create a token and add it to the localStorage
You can use this script
```js
const username = ""; // your username
const password = ""; // your password

const combined = `${username}:${password}`;
const buff = Buffer.from(combined, "utf-8");
const resultToken = buff.toString("base64");

console.log(resultToken);
```

Then try to add items to the cart from the product page (click cart icon).
Then try to increment/decrement the count in the cart (click `-` and `+` buttons on the product card on the main page or the cart view).
Then try to submit the order (the last step in the cart).
Then go to the "Manage Orders" -- you should see all orders.
You should be able to delete order.
You should be able to manage a single order (update status)

If you have any questions please feel free to ask :)

## [Tasks](https://github.com/rolling-scopes-school/aws/blob/main/aws-developer/08_integration_with_sql_database/task.md#tasks)

### Task 8.1

- [x] 1. Fork a copy of [Cart Service template repository](https://github.com/rolling-scopes-school/nodejs-aws-cart-api)
- [x] 2. Use the guide (https://docs.nestjs.com/faq/serverless) to wrap Nest.js application to AWS Lambda, but replace Serverless Framework with AWS CDK to create and deploy your lamda as you already did in [task 3](https://github.com/rolling-scopes-school/aws/blob/main/aws-developer/03_serverless_api/task.md)
- [x] 3. Deploy your code to AWS Lambda

### Task 8.2

- [x] 1. Use AWS Console to create a database instance in RDS with PostgreSQL and default configuration.
![image](https://github.com/slnchn/nodejs-aws-cart-api/assets/40535982/973de5e5-8f98-45be-b9fb-952bae2aa7a4)
- [x] 2. Connect to database instance via a tool called [DBeaver](https://dbeaver.io/download/) or any other similar tools like DataGrip/PgAdmin
![image](https://github.com/slnchn/nodejs-aws-cart-api/assets/40535982/b50d209f-bd5d-4f68-951d-673fa7ac7993)
- [x] 3. Create Cart and CartItem tables
![image](https://github.com/slnchn/nodejs-aws-cart-api/assets/40535982/d14aa362-8574-4b16-8925-89f9948547e6)

### Task 8.3

- [x] 1. Update source code in the application to use PostgreSQL instead of memory storage.
Done. There are a lot of changes, but you can verify that services rely on database instead of in-memory values
- [x] 2. Integrate with RDS
Done -- https://github.com/slnchn/nodejs-aws-cart-api/pull/1/files#diff-089f4f2474b64391c42b6e66aed33977e132058d92108f0a63234a7862e1f8b8R25-R44
Also, all data is stored in the RDS
- [x] 3. Extend your AWS CDK Stack file with credentials to your database instance and pass it to lambda’s environment variables section.
Done -- https://github.com/slnchn/nodejs-aws-cart-api/pull/1/files#diff-ceebe7defe435633e01cc335a33cd0238a9e521072872008f1e5bbe302c6be36R27-R34

### Task 8.4

- [x] 1. Commit all your work to separate branch (e.g. task-8 from the latest master) in your new repository.
- [x] 2. Create a pull request to the master branch.
- [x] 3. Submit link to the pull request to Crosscheck page in [RS App](https://app.rs.school/).

### Additional scope

- [x] +20 (All languages) - Create orders table and integrated with it Order model
Set status to 'ORDERED' after checkout instead of cart deletion.
Done -- there is an [order service](https://github.com/slnchn/nodejs-aws-cart-api/pull/1/files#diff-5826031ecf74a3db1f74cbca740fb5173eef6200a4de41471fbb9238be772287) and [order entity](https://github.com/slnchn/nodejs-aws-cart-api/pull/1/files#diff-4fc288a9e9f1a0761e5889091b9522243ad465bfe6d9c959d9406e26b339bdad)
Orders management also was implemented on the client side https://d1vdv7tmd1pu8.cloudfront.net/
Orders table database screenshot
![image](https://github.com/slnchn/nodejs-aws-cart-api/assets/40535982/14e4cf8d-1654-4959-9e34-aea56597d493)

- [x] +4 (All languages) - Create users table and integrate with it
Done. There is a [user entity](https://github.com/slnchn/nodejs-aws-cart-api/pull/1/files#diff-a7d48eef257cfde1b4373afc0c2b13bfa1c19bcd3b8ed6d3bd294bb64eefffc8).
Also, the [users service](https://github.com/slnchn/nodejs-aws-cart-api/pull/1/files#diff-53c6b9187bd5e88d9a6e04a11b7e98992c3fc87209b8768f7d11e7d7da271c86) uses this table
Also, I can get username from the users table in orders by joining orders and users table by `user_id` 
![image](https://github.com/slnchn/nodejs-aws-cart-api/assets/40535982/cd9a77bc-98ab-4d39-8670-07c28439b1ac)

Users table database screenshot
![image](https://github.com/slnchn/nodejs-aws-cart-api/assets/40535982/b03a8293-fc7e-4e17-9b63-a4a85b5ebc21)

- [x] +3 (All languages) - Transaction based creation of checkout
Done -- https://github.com/slnchn/nodejs-aws-cart-api/pull/1/files#diff-5826031ecf74a3db1f74cbca740fb5173eef6200a4de41471fbb9238be772287R74-R97

- [x] +3 (All languages) - Integrate Cart service with FE repository
Done. Please follow the https://d1vdv7tmd1pu8.cloudfront.net/ and verify that you can
- add product to the cart from the main page
- increment/decrement product count from the main page or cart
- submit order (from cart view) (please note that it also changes cart status to "ORDERED", you can see it in the Network tab responses)
- view orders from the "Manage Orders"
- delete orders (please note that it also changes cart status back to "OPEN", you can see it in the Network tab responses)
- manage single order (change status and track statusHistory)
